### DISCUSSION TITLE
Would ya'll be interested in a professional UX designer pitching in?

### DISCUSSION THREAD

**Discussion by jcklpe** [cgeo/cgeo#10003] [General]
I love c:geo. I've been using it as my only geocaching app for over a decade. 

But after a few years away I came back to it and was like.... oh dang this UI.... leaves a lot to be desired. 

I work fulltime as a UX designer. I can do frontend web development, but wouldn't really be able to help with a Java heavy system. 

But if there were interest in me contributing I'd like to give back to the app that has given me so much personal enjoyment. 

I know how open source projects can be... sometimes its... a little hard to see any change. I know to a lot of devs they can look at pretty old fashioned UI and say "What! Why does this need to change at all?!" but I'm willing to have that conversation and talk through those concerns. I'm just feeling out if there's any serious appetite for this.

**Comment by SammysHP** [comment]
Sure, every help is welcome!

But we already had some drafts in the past. The problem is, that being a designer is not enough. We need someone who takes the new design and translates it into Android XML. Also many functions depend on specific design elements in the code. Not an easy task and that's why we still use the very ancient design.

**Reply by jcklpe** [reply]
I've never written Android XML but I can write HTML and JSX, and what I know of XML is that it's just extensible HTML right? 

If I were to put together a redesign it would approach it from a code agnostic approach first, seeking to produce a design that reflects the mental model of the user and then streamline that experience. (I don't use this exact method but it's similar to Object Oriented UX: https://alistapart.com/article/object-oriented-ux/ ) 

What I'm saying here is that the approach wouldn't use the XML as a starting point. Though like I said I'd be willing to help write it probably. (just took a look at it. How odd, xml uses tags to do styling? Hmm that's interesting! Didn't know that was a thing)

**Reply by jcklpe** [reply]
Just looking over replies, and it def looks like there's already been a few stabs at this taken. Doing design in an open source context is always hard. As the saying goes "every completed design just creates more work for the developers". That's a tough problem to solve.

**Reply by SammysHP** [reply]
Android layout XML has not much in common with HTML and CSS.

**Comment by eddiemuc** [comment]
Being myself a programmer with a very dumb sense for UX, I would LOVE it to have support for UX topics. In my impression we have very many long discussions for UX topics for certain features which I feel would be much shorter if there was some sort of guideline to cling to. Some examples from my personal past:
*  SAF folder migration 
* map attribution display (solved by copying UX from locus)
* Offline log image selection

What we maybe cannot easily do is to change the whole c:geo UX to a new design, this would imho more work than we can reasonably do with current team. But a guideline or road book on how to evolve UX step by step and to use for new features/Dialogs/... would be a great help to me.

**Comment by fm-sys** [comment]
This would be more than welcome. 

However as we are all just working on c:geo in our free time, a redesign should be done in small steps. First of all, maybe replace the general look of buttons, backgrounds etc, and then rework some activities step by step. In addition, it is also important that no functions are lost because I know many apps where a redesign has completely backfired. But there may also be quite some discussion about potential changes - it will not be easy... 😅 
Even if the UI isn't really beautiful at the moment, it's quite effective. In general, it should apply: effectiveness over beauty! 

Here is a (probably incomplete) list of issues in which we had dealt with a redesign:
 #7030
 #5927
 #7587
 #4597
 #4319
 #9373

Also there's a huge topic dealing with refactoring the filters...
See #5877,  #9081 or other issues in https://github.com/cgeo/cgeo/projects/16

**Reply by moving-bits** [reply]
Any help bringing UI/UX a bit forward would be very welcomed, thanks for your offer!

A step by step approach (as already mentioned by @fm-sys) is needed to make it feasible for our small team, as much of the UX directly translates into source code because of UI elements and workflows.

Starting points could be our homescreen or our theming (#9373) to get a consistent look & feel.

**Reply by jcklpe** [reply]
@fm-sys Those are some great links, thank you for sharing. 

That first link https://github.com/cgeo/cgeo/issues/7030  is a pretty good first mock up actually. 

I guess my first question would be: why hasn't that been implemented? So is there a reason why that wasn't used? Is there additional information that needs to be dealt with? How do ya'll want to approach it?

-----
A more general note:

I think it's important to put forth a position about design. Design is not primarily about how things "look" but rather how things are structured. Visual design is part of UX design, in the same way that grammar and syntax are part of writing a novel. 

My day job is primarily working with legacy apps (I work largely for government), and updating them. The is commonly concern about "losing functionality" when doing redesigns. I want to assure you that if I were to address this stuff on a structural level I would seek to do this in spirit, though I also want to make sure that it's understood that all necessary actions should be preserved for a User, they might not always look like the exact same "functionality" from a development perspective. 

And example of this from an old project I worked on. I was working on an application that was used to make appointments. There was a "reschedule appointment" page where the user could select an appointment and transfer it to a new timeslot. There was also a separate "change appointment location" page where a user could select an appointment and move it to a different location. I "trashed" both of those functionalities. Instead a user could click the "edit appointment" button and they could edit the timeslot or location of the appointment. In essence the needs of the User were preserved but the surface level UI way they were address was  consolidated. 

All this said, it does seem like perhaps it might be best to address the "look and feel" stuff first. The #7030 design mocks actually demonstrate a pretty decent example of how improving just the styling can benefit the app. So to loop back around: how can we carry that work forward?

**Reply by fm-sys** [reply]
> That first link #7030 is a pretty good first mock up actually.
>
> I guess my first question would be: why hasn't that been implemented? So is there a reason why that wasn't used? Is there additional information that needs to be dealt with? How do ya'll want to approach it?

The author of that issue was no developer as well, and at this time nobody of the team had the time to step into it.

Also there were some minor problems with the scratch. What if the user is logged in with multiple accounts?  What should happen to the advanced search page? And also the global cache filter would need to be degraded (Which makes sense for sure, but means additional effort) In addition, most users like the dark design (It's now modern again 😉). When we make a redesign, the dark design and the orange accent color should be definitely preserved.

One idea would be to ask our users about a redesign and show them some ideas. (linked to "Make user survey" #9976) 

In addition, I would be willing to help you with the coding part for a redesign if time permits.

I think the first task would be to switch to material design as base theme (more a development task) and then refactor our start screen (Including getting rid of the global cache filter). We will be busy with this for the next time... 😅

**Reply by moving-bits** [reply]
One thing to keep in mind with regard to the draft from 2018: c:geo has changed quite a bit (in some regards) since then. One example: With yesterday's release we introduced an installation wizard which tries to help with the onboarding process. It's not strictly following the draft from 2018, but some of its concepts are implemented. (Installation wizard will pop up automatically after a fresh install of c:geo.)

**Reply by fm-sys** [reply]
Yes, IMHO the installation wizard actually was the first step towards a redesign. The onboarding is nearly perfect from UX perspective, so it's time to continue with other parts...

**Reply by jcklpe** [reply]
I think adhering to material design principles would go a long way to improve the general ui for sure. 

Let me know what I can do to help.

**Reply by moving-bits** [reply]
One thing to start with could be to define a set of font sizes for different use cases, e. g.:
- mainscreen
  - toolbar text
  - icon label text
  - login and coordinates text
  - GPS info
  - notification text
- lists (cache list etc.)
  - primary info text (e. g.: cache name)
  - secondary info text (e. g.: geocode, D/T rating etc.)
  - toolbar primary text (e. g.: cache list name)
  - toolbar secondary text (e. g.: number of caches)
- details page (e. g. cache details)
  - default display text
  - editor text
  - editor title text
  - ...

etc.

to have it well-proportioned to each other and usable.

Examples from today are:
- cache description in cache details page is in `14sp`
- mainscreen icon label are `13sp`
- cachelist: cachename is in `18sp`, secondary text in `12sp`

This we could implement then while we are migrating theming/styling to standard material design components and libraries.

**Reply by jcklpe** [reply]
So a good way to talk about this is in terms of a defining a "type palette" or "type hierarchy" (which in web is usually typed to the h1-h6 semantic markup). Here's an example: 

![image](https://user-images.githubusercontent.com/9451726/108941495-d6ce3e80-761a-11eb-858d-916f8adf7589.png)

I'm more used to working in terms of px, pt, or em in web contexts, but looking up sp I'm guessing that it's essentially the same as pt. (roughly like a pixel but scaled to be independent of the screen resolution). 

My first recommendation here would be to bump up all those font sizes by a fair bit. [Usability studies](https://health.gov/healthliteracyonline/display/section-3-3/) strongly support the use of at least a 16px paragraph reading font. [Another good article on the subject here.](https://www.smashingmagazine.com/2011/10/16-pixels-body-copy-anything-less-costly-mistake/) [And here's a study by MIT.](https://journals.sagepub.com/doi/10.1177/1541931213601698)

So if I were writing out a typical type hierarchy I'd put it as such:

h1: 40 bold
h2: 32
h3: 28
h4: 22
h5: 20
h6: 18
p (paragraph or block text): 16


And then having established that type hierarchy I would then assign different parts of the interface to different levels of that hierarchy. 

As a baseline everything would need to be 16 and above, with maybe very very rare exceptions. 

So going off your examples real quick I'd say:

mainscreen

- toolbar text : p
- icon label text: h5
- login and coordinates text: p
- GPS info: not sure what this is referring to. 
- notification text : not clear on this either. If you mean the badge on the the icons then I'd put that as 16 or perhaps lower. It really depends on the icons though. 

lists (cache list etc.)

- primary info text (e. g.: cache name): h5 though maybe h4
- secondary info text (e. g.: geocode, D/T rating etc.) : p though I would also eliminate some of this information as unnecessary. What is the point of the geocode being in the list? 
- toolbar primary text (e. g.: cache list name): h6
- toolbar secondary text (e. g.: number of caches): p

details page (e. g. cache details)

- default display text : p
- editor text : p
- editor title text : h5


After going through these, it occurs to me that for a "list heavy" app like this, rather than a "reading long passages of text" heavy the use of most of the higher level h3-h1 is unnecessary, but I didn't delete my notes on that just to make sure you had full context on my thought process.

**Reply by jcklpe** [reply]
Any thoughts @moving-bits ?

**Reply by moving-bits** [reply]
Hi @jcklpe,

sorry for not coming back to you on this earlier - we are still pretty busy with our current release (including the major changes needed due to Android changing their storage access behavior), and have been postponed "a bit" by the recent website changes. In parallel we are currently preparing the next release including such things as an integrated offline routing engine and a connector for Lab Adventure caches (and more).

But besides that, there will also be a tiny step forward in terms of c:geo styling, although visible for the developers only yet: With #10049 (included in next main release) we have unified c:geo's use of the base building blocks of every Android app, the so called "Activities", to be based on the very same `AppCompatActivity`. Also we have solved some special theming issues by removing dependency on some device specific themes. Both changes are targeted to make a consistent theming a bit easier.

Next step could be to remove all individual theming and make it be based on the values you have proposed above (big thanks for that!). Alas, as we have so many different themes and styles in c:geo currently, this will be a major task, probably requiring weeks or even months to finish in all details. A good starting point for this will be after branching the next release.

Again, a big "thank you" for offering us to help with styling, layout and graphics. This is highly appreciated!

**Reply by jcklpe** [reply]
@moving-bits Sounds good! I'm pitching in some help on the icon right now. If you're interested I'd love to hear what c:geo means to you, as a visual brand :) https://github.com/cgeo/cgeo/issues/10522

**Reply by fm-sys** [reply]
> My first recommendation here would be to bump up all those font sizes by a fair bit. [Usability studies](https://health.gov/healthliteracyonline/display/section-3-3/) strongly support the use of at least a 16px paragraph reading font. [Another good article on the subject here.](https://www.smashingmagazine.com/2011/10/16-pixels-body-copy-anything-less-costly-mistake/) [And here's a study by MIT.](https://journals.sagepub.com/doi/10.1177/1541931213601698)

I think this would be a good thing. Probably, even material design would do this automatically for us. We just overwrite the font size at many places currently...

**Comment by Lineflyer** [comment]
Sure, help is welcome. The design needs a big overhaul.. 
A difficult task here is also to preserve a bit the look and feel IMHO so that users still  recognize their c:geo while they at the same time benefit from better usability and style.

**Comment by bekuno** [comment]
@jcklpe 
Are you also a graphic designer? We need also support in this are. 
Currently we want to reduce the need of rastered graphics and therefore move to vector graphics.
But our nice logo design does not good convert to a android vector graphic. For details see #10468.

**Reply by jcklpe** [reply]
@bekuno   : Yup, I can also do graphic design. I'll post some questions to that ticket.

**Comment by eddiemuc** [comment]
In search of a base widget to use for a slider for filtering views (things like filtering by terrain, difficuntly, size, ... as wished in #9081) I came across the Sliders of material.io: https://material.io/components/sliders 

To my surprise the material.io libraries are already a dependency for c:geo code, although I do not know why (we seem not to use any of their components yet - does anyone know why this was included?). But this got me thinking:

Would it be a valid approach from technical side to base a new c:geo design on the principles and methods/components offered on material.io? 
* They somehow seem to provide exactly this "link" between great GUI design and technical components to easily implement it. 
* They seem to both allow maximum freedom wrt designing app-specific layouts while providing a solid technical implementation to realize them into code and to change design later without impacting code too much.
* They seem to reuse Androids SDK wherever possible (e.g. for buttons they seem to "just" provide styles for the Android default button classes) and add widgets/components only where the Android SDK does not offer anything itself (e.g. Sliders) - which is a good approach imho.

Also we are alread heavily using many icons from this project anyway.

This would NOT mean that we do not need an "own" design for c:geo - quite the contratry! I just think this could give us both a starting point and a technical base to implement such a design in a clean, future-stable way.

What do you think?

More info: 
* Overview: https://material.io
* Design: https://material.io/design
* Components: https://material.io/components

**Reply by fm-sys** [reply]
I also think that there is a lot of potential in it! So the first attempt could actually be to just change the base theme to material design. And then adjust the places where we have some old fashioned selfmade solutions...

**Reply by jcklpe** [reply]
Yeah I think that material design provides a pretty solid starting point for modernizing c:geo
